### PR TITLE
Add properties to material

### PR DESCRIPTION
**Pull request by sannecottaar**
Here is a version of using properties in Material (and thus in Mineral, Solid Solution, and Composite). 

Some advantages: 
- Nothing unused gets calculated in set_state. 
- There is consistency in the properties now (in the previous case some where functions and some where calculated in set_state). 

Downside:
- Currently everything gets recalculated every time. The upside if this is that you can change parameters of the material, and the properties change without setting the state again. The downside is of course the recalculations of something that might have been calculated before. Ian has suggested caching the properties, but other opinions and suggestions are welcome. 

Lots of routines (mostly in main.py) are redundant now. I have not adapted all the examples to use the new evaluate functions yet (which just spits out lists of requested properties over given ranges of P and T). Awaiting your comments and ideas first.

### CODE REVIEW

**Comment by sannecottaar** [comment]
Now all the properties are reset during set_state. This speeds up example_grid.py with a 100x100 grid in P and T by a factor of 2.

File: burnman/mineral.py [hunk 1]
```diff
@@ -121,135 +128,193 @@ def set_state(self, pressure, temperature):
...
     def molar_entropy(self):
         """
         Returns enthalpy of the mineral [J]
         """
-        return self.S
+        if self._molar_entropy is None :
+            self._molar_entropy = self.method.entropy(self.pressure, self.temperature, self.molar_volume, self.params)
+        return self._molar_entropy
+
+    gibbs = molar_gibbs
```
Comment (tjhei) [inline burnman/mineral.py]: should we remove those?

File: burnman/mineral.py [hunk 2]
```diff
@@ -43,7 +43,14 @@ def __init__(self):
         self.method = None
         if 'equation_of_state' in self.params:
             self.set_method(self.params['equation_of_state'])
-
+        if 'name' in self.params:
```
Comment (tjhei) [inline burnman/mineral.py]: Can we create Material.reset() that is called from here and the other classes?

File: burnman/mineral.py [hunk 3]
```diff
@@ -107,149 +114,227 @@ def set_state(self, pressure, temperature):
...
-            if self.pressure == pressure and self.temperature == temperature and self.old_params == self.params:
-                return
-        except AttributeError:
-            pass  #do nothing
-
-        self.pressure = pressure
-        self.temperature = temperature
-        self.old_params = self.params
+        self._pressure = pressure
+        self._temperature = temperature
```
Comment (tjhei) [inline burnman/mineral.py]: can we call Material.set_state() here instead?

File: burnman/mineral.py [hunk 4]
```diff
@@ -107,149 +114,227 @@ def set_state(self, pressure, temperature):
...
-        try:
-            self.S = self.method.entropy(self.pressure, self.temperature, self.V, self.params)
-        except (KeyError, NotImplementedError):
-            self.S = float('nan')
-        try:
-            self.H = self.method.enthalpy(self.pressure, self.temperature, self.V, self.params)
-        except (KeyError, NotImplementedError):
-            self.H = float('nan')
+        #reset properties
+        class_items = Mineral.__dict__.iteritems()
```
Comment (tjhei) [inline burnman/mineral.py]: call reset() here

File: burnman/solidsolution.py [hunk 5]
```diff
@@ -98,60 +117,294 @@ def set_composition(self, molar_fractions ):
...
-        self.H = sum([ self.endmembers[i][0].H * self.molar_fractions[i] for i in range(self.n_endmembers) ]) + self.excess_enthalpy
-        self.S = sum([ self.endmembers[i][0].S * self.molar_fractions[i] for i in range(self.n_endmembers) ]) + self.excess_entropy
-        self.V = sum([ self.endmembers[i][0].V * self.molar_fractions[i] for i in range(self.n_endmembers) ]) + self.excess_volume
-        self.C_p = sum([ self.endmembers[i][0].C_p * self.molar_fractions[i] for i in range(self.n_endmembers) ])
-        self.alpha = (1./self.V) * sum([ self.endmembers[i][0].alpha * self.endmembers[i][0].V * self.molar_fractions[i] for i in range(self.n_endmembers) ])
-        self.K_T = self.V * 1./(sum([ self.endmembers[i][0].V / (self.endmembers[i][0].K_T)  * self.molar_fractions[i] for i in range(self.n_endmembers) ]))
+
+
+
+    def calcgibbs(self, pressure, temperature, molar_fractions):
```
Comment (tjhei) [inline burnman/solidsolution.py]: delete

File: burnman/solidsolution.py [hunk 6]
```diff
@@ -98,60 +117,294 @@ def set_composition(self, molar_fractions ):
...
+                if element not in bulk_composition:
+                    bulk_composition[element] = self.molar_fractions[i]*mbr_composition[element]
+                else:
+                    bulk_composition[element] += self.molar_fractions[i]*mbr_composition[element]
+        return bulk_composition
+                    
     def set_method(self, method):
         for i in range(self.n_endmembers):
             self.endmembers[i][0].set_method(method)
         self.method = self.endmembers[0][0].method
```
Comment (tjhei) [inline burnman/solidsolution.py]: call self.reset()

File: burnman/mineral.py [hunk 7]
```diff
@@ -110,149 +118,203 @@ def set_state(self, pressure, temperature):
...
-            self.helmholtz = float('nan')
-        try:
-            self.S = self.method.entropy(self.pressure, self.temperature, self.V, self.params)
-        except (KeyError, NotImplementedError):
-            self.S = float('nan')
-        try:
-            self.H = self.method.enthalpy(self.pressure, self.temperature, self.V, self.params)
-        except (KeyError, NotImplementedError):
-            self.H = float('nan')
+            raise AttributeError, "no method set for mineral, or equation_of_state given in mineral.params"
```
Comment (tjhei) [inline burnman/mineral.py]: not python 3 ("raise AttributeError, ...")

File: burnman/eos/slb.py [hunk 8]
```diff
@@ -89,26 +116,48 @@ def volume(self, pressure, temperature, params):
         a2_iikk = -12.*params['grueneisen_0']+36.*pow(params['grueneisen_0'],2.) - 18.*params['q_0']*params['grueneisen_0'] # EQ 47
 
         b_iikk= 9.*params['K_0'] # EQ 28
-        b_iikkmm= 27.*params['K_0']*(params['Kprime_0']-4.) # EQ 29
+        b_iikkmm= 27.*params['K_0']*(params['Kprime_0']-4.) # EQ 29z
+        #f = lambda x: 0.5*(pow(V_0/x,2./3.)-1.) # EQ 24
+        #debye_T = lambda x : self.__debye_temperature(params['V_0']/x, params)
+        
+        debye_temperature = lambda x : Debye_0 * np.sqrt(1. + a1_ii * 0.5*(pow(V_0/x,2./3.)-1.) + 1./2. * a2_iikk*0.5*(pow(V_0/x,2./3.)-1.)*0.5*(pow(V_0/x,2./3.)-1.))
+        E_th =  lambda x : debye.thermal_energy(temperature, debye_temperature(x),  n) #thermal energy at temperature T
```
Comment (ian-r-rose) [inline burnman/eos/slb.py]: This seems inadvertently reintroduced.

File: burnman/eos/slb.py [hunk 9]
```diff
@@ -28,6 +34,15 @@ def _debye_temperature(self,x,params):
         a1_ii = 6. * params['grueneisen_0'] # EQ 47
         a2_iikk = -12.*params['grueneisen_0']+36.*pow(params['grueneisen_0'],2.) - 18.*params['q_0']*params['grueneisen_0'] # EQ 47
         return params['Debye_0'] * np.sqrt(1. + a1_ii * f + 1./2. * a2_iikk*f*f)
+ 
+    def __debye_temperature_shortcut(self,x,Debye_0,a1_ii,a2_iikk):
```
Comment (ian-r-rose) [inline burnman/eos/slb.py]: Remove the dunderscore

File: burnman/material.py [hunk 10]
```diff
@@ -100,20 +126,507 @@ def unroll(self):
...
+        
+        Returns
+        -------
+        heat_capacity_p : float
+        """
+        raise NotImplementedError("need to implement heat_capacity_p() in derived class!")
+        return None
+
+
+    def copy(self):
```
Comment (tjhei) [inline burnman/material.py]: remove?

File: burnman/eos/slb.py [hunk 11]
```diff
@@ -89,26 +116,48 @@ def volume(self, pressure, temperature, params):
...
             tol = 0.0001
-            sol = opt.fmin(lambda x : pow(self._volume(x,pressure,temperature,V_0,T_0,Debye_0,n,a1_ii,a2_iikk,b_iikk,b_iikkmm),2.0), 1.0*params['V_0'], ftol=tol, full_output=1, disp=0)
+            sol = opt.fmin(lambda x : self.__volume(x,args)*self.__volume(x,args), 1.0*params['V_0'], ftol=tol, full_output=1, disp=0)
             if sol[1] > tol*2:
                 raise ValueError('Cannot find volume, likely outside of the range of validity for EOS')
             else:
                 warnings.warn("May be outside the range of validity for EOS")
                 return sol[0][0]
 
+    def grueneisen_parameter_shortcut(self, pressure, temperature, volume, params,a1_ii,a2_iikk):
```
Comment (ian-r-rose) [inline burnman/eos/slb.py]: Underscore, since it is internal?

File: burnman/eos/slb.py [hunk 12]
```diff
@@ -63,6 +78,18 @@ def _thermal_pressure(self,T,V, params):
         gr = self.grueneisen_parameter(0., T, V, params) # P not important
         P_th = gr * debye.thermal_energy(T,Debye_T, params['n'])/V
         return P_th
+    
+    def __volume(self,x,pressure,temperature,V_0,T_0,Debye_0,n,a1_ii,a2_iikk,b_iikk,b_iikkmm):
```
Comment (ian-r-rose) [inline burnman/eos/slb.py]: Dunderscore

File: burnman/material.py [hunk 13]
```diff
@@ -100,20 +126,507 @@ def unroll(self):
...
+        vars_list : list of strings
+        Variables to be returned for given conditions
+        pressure : array of float
+        Array of pressures in [Pa].
+        temperature : float
+        Array of temperaturesin [K].
+        
+        Returns
+        -------
+
```
Comment (tjhei) [inline burnman/material.py]: missing comment what it returns

File: burnman/material.py [hunk 14]
```diff
@@ -77,8 +96,15 @@ def set_state(self, pressure, temperature):
...
         """
-        self.pressure = pressure
-        self.temperature = temperature
+        self.reset()
+        
+        self._pressure = pressure
+        self._temperature = temperature
+    
+    def reset(self):
+        self._cached = {}
```
Comment (tjhei) [inline burnman/material.py]: should reset also set _temperature/_pressure to None?

File: burnman/main.py [hunk 15]
```diff
@@ -12,6 +12,7 @@
 
 class ElasticProperties(object):
     """
+    REDUNDANT
```
Comment (tjhei) [inline burnman/main.py]: what is with all these comments in main.py? Should they be TODOs?

File: burnman/eos/slb.py [hunk 16]
```diff
@@ -89,26 +116,48 @@ def volume(self, pressure, temperature, params):
...
         # conservative guess:
         a = 0.6*params['V_0']
         b = 1.2*params['V_0']
+        args=pressure,temperature,V_0,T_0,Debye_0,n,a1_ii,a2_iikk,b_iikk,b_iikkmm
         # if we have a sign change, we are done:
-        vol_a =self._volume(a,pressure,temperature,V_0,T_0,Debye_0,n,a1_ii,a2_iikk,b_iikk,b_iikkmm)
-        vol_b =self._volume(b,pressure,temperature,V_0,T_0,Debye_0,n,a1_ii,a2_iikk,b_iikk,b_iikkmm)
-        if vol_a*vol_b<0:
-            return opt.brentq(self._volume, a, b,args=(pressure,temperature,V_0,T_0,Debye_0,n,a1_ii,a2_iikk,b_iikk,b_iikkmm))
+        if self.__volume(a,pressure,temperature,V_0,T_0,Debye_0,n,a1_ii,a2_iikk,b_iikk,b_iikkmm)*self.__volume(b,pressure,temperature,V_0,T_0,Debye_0,n,a1_ii,a2_iikk,b_iikk,b_iikkmm)<0:
```
Comment (tjhei) [inline burnman/eos/slb.py]: please fix

File: burnman/averaging_schemes.py [hunk 17]
```diff
@@ -64,8 +64,7 @@ def average_shear_moduli(self, volumes, bulk_moduli, shear_moduli):
     def average_density(self, volumes, densities):
         """
         Average the densities of a composite, given a list of volume
-        fractions and densitites. This is the only method of
-        :class:`burnman.averaging_schemes.AveragingScheme` which is implemented in the base class,
+        fractions and densitites. This  is implemented in the base class,
```
Comment (tjhei) [inline burnman/averaging_schemes.py]: fix double space

File: burnman/composite.py [hunk 18]
```diff
@@ -6,22 +6,26 @@
 import numpy as np
 import warnings
 
-from .material import Material
+from .material import Material, material_property
 from .mineral import Mineral
 
-def check_pairs(fractions, minerals):
+from burnman import averaging_schemes
+from burnman import chemicalpotentials
```
Comment (tjhei) [inline burnman/composite.py]: fix imports

File: burnman/composite.py [hunk 19]
```diff
@@ -6,22 +6,26 @@
 import numpy as np
 import warnings
 
-from .material import Material
+from .material import Material, material_property
 from .mineral import Mineral
 
-def check_pairs(fractions, minerals):
+from burnman import averaging_schemes
```
Comment (ian-r-rose) [inline burnman/composite.py]: These should be 
```
from . import averaging_schemes
from . import chemicalpotentials
```

File: burnman/composite.py [hunk 20]
```diff
@@ -33,63 +37,144 @@ class Composite(Material):
...
         Create a composite using a list of phases and their fractions (adding to 1.0).
 
         Parameters
         ----------
-        fractions: list of floats
-            molar fraction for each phase.
         phases: list of :class:`burnman.Material`
             list of phases.
+        fractions: list of floats
+            molar fraction for each phase.
```
Comment (tjhei) [inline burnman/composite.py]: document fraction_type

File: burnman/material.py [hunk 21]
```diff
@@ -2,6 +2,23 @@
 # This file is part of BurnMan - a thermoelastic and thermodynamic toolkit for the Earth and Planetary Sciences
 # Copyright (C) 2012 - 2015 by the BurnMan team, released under the GNU GPL v2 or later.
 
+import numpy as np
+import copy
```
Comment (ian-r-rose) [inline burnman/material.py]: Necessary import?

File: tests/test_endmembers.py [hunk 22]
```diff
@@ -60,21 +60,23 @@ class test_endmembers(BurnManTest):
     def test0_T_ref(self):
         fo=forsterite()
         fo.params['P_0'] = 0.99999999999e5
-
+
```
Comment (ian-r-rose) [inline tests/test_endmembers.py]: whitespace

File: burnman/composite.py [hunk 23]
```diff
@@ -33,63 +37,144 @@ class Composite(Material):
...
+                self.phase_names.append('')
+
+        if fractions is not None:
+            self.set_fractions(fractions, fraction_type)
+        else:
+            self.molar_fractions=None
+
+        self.set_averaging_scheme('VoigtReussHill')
+
+        self._pressure = None
```
Comment (tjhei) [inline burnman/composite.py]: can we call ```Material.__init__(self)``` instead?

File: burnman/composite.py [hunk 24]
```diff
@@ -33,63 +37,144 @@ class Composite(Material):
...
+                self.phase_names.append(self.phases[i].name)
+            except AttributeError:
+                self.phase_names.append('')
+
+        if fractions is not None:
+            self.set_fractions(fractions, fraction_type)
+        else:
+            self.molar_fractions=None
+
+        self.set_averaging_scheme('VoigtReussHill')
```
Comment (tjhei) [inline burnman/composite.py]: why do we set VRH now?

File: burnman/averaging_schemes.py [hunk 25]
```diff
@@ -64,8 +64,7 @@ def average_shear_moduli(self, volumes, bulk_moduli, shear_moduli):
     def average_density(self, volumes, densities):
         """
         Average the densities of a composite, given a list of volume
-        fractions and densitites. This is the only method of
-        :class:`burnman.averaging_schemes.AveragingScheme` which is implemented in the base class,
```
Comment (ian-r-rose) [inline burnman/averaging_schemes.py]: Is there a reason for this documentation change?

File: burnman/composite.py [hunk 26]
```diff
@@ -33,63 +37,144 @@ class Composite(Material):
...
-
-        assert(len(phases)==len(fractions))
+        
         assert(len(phases)>0)
+        self.phases = phases
+        
+        self.phase_names = []
+        for i in range(len(self.phases)):
+            try:
+                self.phase_names.append(self.phases[i].name)
```
Comment (tjhei) [inline burnman/composite.py]: I don't like how names are handled with a try/except. We used to have to_string() which always works.

File: burnman/composite.py [hunk 27]
```diff
@@ -33,63 +37,144 @@ class Composite(Material):
...
+        for i, composition in enumerate(phase_compositions):
+            if composition != []:
+                self.phases[i].set_composition(composition)
+            
 
     def set_method(self, method):
         """
         set the same equation of state method for all the phases in the composite
         """
-        for (fraction, phase) in self.children:
```
Comment (tjhei) [inline burnman/composite.py]: do we need to call reset here?

File: burnman/composite.py (hunk 20 above)
Comment (tjhei) [inline burnman/composite.py]: also the documentation of "fraction" is no longer correct (depends on fraction_type)

File: burnman/composite.py [hunk 28]
```diff
@@ -33,63 +37,144 @@ class Composite(Material):
...
-        print("%sComposite:" % indent)
-        indent += "  "
-        for (fraction, phase) in self.children:
-            print("%s%g of" % (indent, fraction))
-            phase.debug_print(indent + "  ")
+        if fraction_type == 'molar':
+            molar_fractions = fractions
+        elif fraction_type == 'mass':
+            molar_fractions = self._mass_to_molar_fractions(self.phases, fractions)
+        elif fraction_type == 'volume':
```
Comment (tjhei) [inline burnman/composite.py]: we need to have a test for this

File: burnman/composite.py [hunk 29]
```diff
@@ -33,63 +37,144 @@ class Composite(Material):
...
+                    bulk_composition[element] = self.molar_fractions[i]*composition[element]
+                else:
+                    bulk_composition[element] += self.molar_fractions[i]*composition[element]
+        return bulk_composition
+            
+
+    def debug_print(self, indent=""):
+        print("%sComposite:" % indent)
+        indent += "  "
+        try: # if fractions have been defined
```
Comment (tjhei) [inline burnman/composite.py]: use ```self.molar_fractions is None``` instead of try/except

File: burnman/composite.py [hunk 30]
```diff
@@ -33,63 +37,144 @@ class Composite(Material):
...
-            minerals.extend(p_min)
-        return (fractions, minerals)
+        try:
+            for i, phase in enumerate(self.phases):
+                p_mineral, p_fraction = phase.unroll()
+                check_pairs(p_mineral, p_fraction)
+                fractions.extend([f*self.molar_fractions[i] for f in p_fraction])
+                phases.extend(p_mineral)
+        except:
+            raise Exception("Unroll only works if the composite has defined fractions.")
```
Comment (tjhei) [inline burnman/composite.py]: use ```self.molar_fractions is None``` instead of try/except

File: burnman/composite.py [hunk 31]
```diff
@@ -97,20 +182,310 @@ def to_string(self):
...
+    def shear_modulus(self):
+        """
+        Returns shear modulus of the mineral [Pa]
+        Aliased with self.G
+        """
+        V_frac = np.array([phase.molar_volume*molar_fraction for (phase, molar_fraction) in zip(self.phases, self.molar_fractions)])
+        K_ph = np.array([phase.adiabatic_bulk_modulus for phase in self.phases])
+        G_ph = np.array([phase.shear_modulus for phase in self.phases])
+        
+        self._shear_modulus = self.averaging_scheme.average_shear_moduli(V_frac, K_ph, G_ph)
```
Comment (tjhei) [inline burnman/composite.py]: please remove all ```self._bla = ``` lines and just return the new value

File: burnman/composite.py [hunk 32]
```diff
@@ -97,20 +182,310 @@ def to_string(self):
...
+        return self._heat_capacity_p
+
+
+
+
+
+    def _calculate_moduli(self):
+        """
+        REWRITE to evaluate endmembers????
+        Calculate the elastic moduli and densities of the individual phases in the assemblage.
```
Comment (tjhei) [inline burnman/composite.py]: this comment seems to be out of date. What is the status of that function?

File: burnman/eos/slb.py [hunk 33]
```diff
@@ -11,6 +11,12 @@
 from . import debye
 from . import equation_of_state as eos
 
+try:
+    from numba import jit
```
Comment (tjhei) [inline burnman/eos/slb.py]: this jit thing doesn't seem to be used

File: burnman/material.py (hunk 14 above)
Comment (tjhei) [inline burnman/material.py]: please document the function. Do we need to make it private (_)?

File: burnman/material.py [hunk 34]
```diff
@@ -100,20 +126,507 @@ def unroll(self):
...
+        
+            
+        Parameters
+        ----------
+        vars_list : list of strings
+        Variables to be returned for given conditions
+        pressure : array of float
+        Array of pressures in [Pa].
+        temperature : float
+        Array of temperaturesin [K].
```
Comment (tjhei) [inline burnman/material.py]: missing space after "temperatures"

File: burnman/mineral.py [hunk 35]
```diff
@@ -45,6 +46,13 @@ def __init__(self):
         self.method = None
         if 'equation_of_state' in self.params:
             self.set_method(self.params['equation_of_state'])
+        if 'name' in self.params:
```
Comment (tjhei) [inline burnman/mineral.py]: I don't like how "name" is handled and only sometimes defined

File: burnman/mineral.py [hunk 36]
```diff
@@ -110,149 +118,203 @@ def set_state(self, pressure, temperature):
...
-            self.H = float('nan')
+             raise AttributeError("no method set for mineral, or equation_of_state given in mineral.params")
+
 
 
-    # The following gibbs function avoids having to calculate a bunch of unnecessary parameters over P-T space. This will be useful for gibbs minimisation.
-    def calcgibbs(self, pressure, temperature):
-        return self.method.gibbs_free_energy(pressure, temperature, self.params)
+    def composition(self):
+        return self.params['formula']
```
Comment (tjhei) [inline burnman/mineral.py]: do we need this function, is this name good, can we have documentation?

File: burnman/mineral.py [hunk 37]
```diff
@@ -110,149 +118,203 @@ def set_state(self, pressure, temperature):
...
+    
+    @material_property
+    def internal_energy(self):
+        """
+        Returns internal energy of the mineral [J]
+        Aliased with self.energy
+        """
+        self._internal_energy = self.method.internal_energy( self, pressure, temperature, volume, params )
```
Comment (tjhei) [inline burnman/mineral.py]: please remove ```self._bla = ``` and just return

File: burnman/mineral.py [hunk 38]
```diff
@@ -110,149 +118,203 @@ def set_state(self, pressure, temperature):
...
+        self._molar_helmholtz = self.method.helmholtz_free_energy(self.pressure, self.temperature, self.molar_volume, self.params)
+        return self._molar_helmholtz
+    
+    @material_property
     def molar_mass(self):
         """
         Returns molar mass of the mineral [kg/mol]
         """
-        return self.params['molar_mass']
+        self._molar_mass = self.params['molar_mass']
```
Comment (tjhei) [inline burnman/mineral.py]: this could be a ```property``` instead of ```material_property``` right?

File: burnman/solidsolution.py [hunk 39]
```diff
@@ -74,13 +75,31 @@ def __init__(self, molar_fractions=None):
         # Number of endmembers in the solid solution
         self.n_endmembers = len(self.endmembers)
 
+        # Endmember names
+        self.endmember_names = []
+        for i in range(self.n_endmembers):
+            self.endmembers[i][0].set_method(self.endmembers[i][0].params['equation_of_state'])
+            try:
+                self.endmember_names.append(self.endmembers[i][0].params['name'])
```
Comment (tjhei) [inline burnman/solidsolution.py]: we should rework name handling

File: examples/example_composite.py [hunk 40]
```diff
@@ -0,0 +1,145 @@
...
+import os, sys, numpy as np, matplotlib.pyplot as plt
+#hack to allow scripts to be placed in subdirectories next to burnman:
+if not os.path.exists('burnman') and os.path.exists('../burnman'):
+    sys.path.insert(1,os.path.abspath('..'))
+
+import burnman
+from burnman import minerals
+
+if __name__ == "__main__":
+    '''
```
Comment (tjhei) [inline examples/example_composite.py]: weird comment formatting

File: examples/example_composite.py [hunk 41]
```diff
@@ -0,0 +1,145 @@
...
+            self.name='simple_mantle'
+            phases = [forsterite, enstatite]
+            burnman.Composite.__init__(self, phases)
+
+    assemblage=simple_mantle()
+
+    assemblage.set_fractions([0.8, 0.2])
+    print assemblage.composition()
+    assemblage.set_state(P, T)
+    print 'Assemblage 1 (endmembers only)'
```
Comment (tjhei) [inline examples/example_composite.py]: this looks like python2 only

File: tests/test_endmembers.py [hunk 42]
```diff
@@ -60,21 +60,23 @@ class test_endmembers(BurnManTest):
...
         T_0 = 2000.
         fo.set_state(1.e5, T_0)
-        fo.params['T_0'] = T_0
-        fo.params['H_0'] = fo.H
-        fo.params['S_0'] = fo.S
-        fo.params['V_0'] = fo.V
-        fo.params['K_0'] = fo.K_T 
-        fo.params['Kdprime_0'] = -fo.params['Kprime_0']/fo.params['K_0']
-        fo.params['a_0'] = fo.alpha
+        fo2=fo.copy()
```
Comment (tjhei) [inline tests/test_endmembers.py]: do we still need to make the copy here?

File: burnman/main.py (hunk 15 above)
Comment (sannecottaar) [inline burnman/main.py]: Depends on what we want...

File: tests/test_endmembers.py (hunk 42 above)
Comment (sannecottaar) [inline tests/test_endmembers.py]: Yes we do. The params are changed while calling these instances for the first time! We can't do this.

**Pull request merged**